### PR TITLE
Annotate phot plot with spec epoch and hover info

### PR DESCRIPTION
**Pull request by scizen9**
This is a draft PR to address [fritz-beta-feedback issue #84](https://github.com/fritz-marshal/fritz-beta-feedback/issues/84).

TODO:
1. pick color scheme to avoid confusion with first/last detections
2. test with multiple spectra

### CODE REVIEW

**Comment by scizen9** [comment]
Have tested with multiple spectra and set line alpha very low (0.1), which does not overwhelm phot plot, but is still visible.

What I don't know how to handle is the fact that each spectrum produces its own hover tool.  It would be great to gang them under one hover tool, if possible.

**Comment by scizen9** [comment]
Now gather all spectral data together into one plot using text glyph ("S").  This produces only one hover tool for all spectra.

**Comment by dannygoldstein** [comment]
@scizen9 can you fix the diff here? Looks like it includes a lot of unrelated commits

**Comment by scizen9** [comment]
> @scizen9 can you fix the diff here? Looks like it includes a lot of unrelated commits

How do I do this?  Would I update the branch from upstream master?

**Comment by scizen9** [comment]
OK, that worked.  Just for posterity, here's what I did:

git rebase master
git push origin annotate-phot-with-spec --force

**Review by kmshin1397 — APPROVED** [review]
Looks good to me! I just realized that the period tab could use some love with adjusting the layout for mobile but I think that's outside the scope of this PR (and probably easiest for me to tackle regardless, once this goes in).

File: skyportal/plot.py [hunk 1]
```diff
@@ -429,6 +440,61 @@ def photometry_plot(obj_id, user, width=600, height=300, device="browser"):
...
+                    data=dict(
+                        s_mjd=s_mjd,
+                        s_y=s_y,
+                        s_date=s_date,
+                        s_tel=s_tel,
+                        s_inst=s_inst,
+                        s_text=s_text,
+                    )
+                ),
+            )
```
Comment (dannygoldstein) [inline skyportal/plot.py:484]: Condense to a single block.

File: skyportal/plot.py [hunk 2]
```diff
@@ -511,6 +577,60 @@ def photometry_plot(obj_id, user, width=600, height=300, device="browser"):
...
+                    s_mjd=s_mjd,
+                    s_y=s_y,
+                    s_date=s_date,
+                    s_tel=s_tel,
+                    s_inst=s_inst,
+                    s_text=s_text,
+                )
+            ),
+        )
+        plot.add_tools(
```
Comment (dannygoldstein) [inline skyportal/plot.py:621]: Condense to a single block.

Comment (scizen9) [reply skyportal/plot.py:621]: I don't understand exactly what you mean.  These are on two separate plots: Mag and Flux.  Do you mean make a function that I can call from both?  That would be one way to combine these.  I would suggest that the more explicit your recommendations are, the better.  Don't assume I am as experienced as you are.

Comment (scizen9) [reply skyportal/plot.py]: So I did make it a function and it works well.  BTW, we could probably do the same thing with the marking of the first and last detections.

File: skyportal/plot.py (hunk 1 above)
Comment (scizen9) [reply skyportal/plot.py]: see resolved conversation below.

File: skyportal/plot.py [hunk 3]
```diff
@@ -162,6 +162,72 @@ def get_color(bandpass_name):
...
+    lower: float
+    upper: float
+        Plot limits allowing calculation of annotation symbol y value
+    """
+    s_mjd = []
+    s_y = []
+    s_date = []
+    s_tel = []
+    s_inst = []
+    s_text = []
```
Comment (dannygoldstein) [inline skyportal/plot.py:182]:
```suggestion
```

File: skyportal/plot.py [hunk 4]
```diff
@@ -162,6 +162,72 @@ def get_color(bandpass_name):
...
+            s.observed_at.hour,
+            s.observed_at.minute,
+            s.observed_at.second,
+        )
+        s_mjd.append(Time(date).mjd)
+        s_y.append(text_y)
+        s_date.append(date)
+        s_tel.append(s.instrument.telescope.name)
+        s_inst.append(s.instrument.name)
+        s_text.append("S")
```
Comment (dannygoldstein) [inline skyportal/plot.py:198]:
```suggestion
```

File: skyportal/plot.py [hunk 5]
```diff
@@ -162,6 +162,72 @@ def get_color(bandpass_name):
...
+    text_y = upper - (upper - lower) * 0.05
+    for s in spectra:
+        date = '%d-%d-%dT%02d:%02d:%02d' % (
+            s.observed_at.year,
+            s.observed_at.month,
+            s.observed_at.day,
+            s.observed_at.hour,
+            s.observed_at.minute,
+            s.observed_at.second,
+        )
```
Comment (dannygoldstein) [inline skyportal/plot.py:192]:
```suggestion
```

File: skyportal/plot.py [hunk 6]
```diff
@@ -162,6 +162,72 @@ def get_color(bandpass_name):
...
+            text_alpha=0.3,
+            text_align='center',
+            source=ColumnDataSource(
+                data=dict(
+                    s_mjd=s_mjd,
+                    s_y=s_y,
+                    s_date=s_date,
+                    s_tel=s_tel,
+                    s_inst=s_inst,
+                    s_text=s_text,
```
Comment (dannygoldstein) [inline skyportal/plot.py:213]:
```suggestion
                    s_mjd=[Time(s.observed_at, format='datetime').mjd for s in spectra],
                    s_y=[text_y] * len(spectra),
                    s_date=[s.observed_at.isoformat() for s in spectra],
                    s_tel=[s.instrument.telescope.name for s in spectra],
                    s_inst=[s.instrument.name for s in spectra],
                    s_text=['S'] * len(spectra),
```

File: skyportal/plot.py [hunk 7]
```diff
@@ -194,6 +260,18 @@ def photometry_plot(obj_id, user, width=600, height=300, device="browser"):
...
+        DBSession()
+        .query(Spectrum)
+        .join(Obj)
+        .join(GroupSpectrum)
+        .filter(
+            Spectrum.obj_id == obj_id,
+            GroupSpectrum.group_id.in_([g.id for g in user.accessible_groups]),
+        )
+    ).all()
+
```
Comment (dannygoldstein) [inline skyportal/plot.py:274]:
```suggestion
spectra = Spectrum.query_records_accessible_by(user).filter(Spectrum.obj_id == obj_id).all()
```

**Review by dannygoldstein — CHANGES_REQUESTED** [review]

File: skyportal/plot.py (hunk 5 above)
Comment (scizen9) [reply skyportal/plot.py:192]: Done

File: skyportal/plot.py (hunk 6 above)
Comment (scizen9) [reply skyportal/plot.py:213]: Done, minus the extraneous commas and organized by calculated and spectral data.

File: skyportal/plot.py (hunk 7 above)
Comment (scizen9) [reply skyportal/plot.py:274]: Done!

File: skyportal/plot.py (hunk 3 above)
Comment (scizen9) [reply skyportal/plot.py]: Done.

File: skyportal/plot.py (hunk 4 above)
Comment (scizen9) [reply skyportal/plot.py]: Done.

**Review by dannygoldstein — APPROVED** [review]

**Comment by dannygoldstein** [comment]
will merge on green

**Comment by stefanv** [comment]
I pushed some small docstring fixes.

**Pull request merged**